src/Compilers/CSharp/Test/Semantic/Semantics/StaticNullChecking_FlowAnalysis.cs

```diff
@@ -0,0 +1,1048 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using Microsoft.CodeAnalysis.CSharp.Test.Utilities;
+using Xunit;
+
+namespace Microsoft.CodeAnalysis.CSharp.UnitTests.Nullable
+{
+    public class StaticNullChecking_FlowAnalysis : CSharpTestBase
+    {
+        [Fact]
+        public void ImplicitlyTypedArrayCreation_01()
+        {
+            var source =
+@"class C
+{
+    static void F(object x, object? y)
+    {
+        var a = new[] { x, x };
+        a[0].ToString();
+        var b = new[] { x, y };
+        b[0].ToString();
+        var c = new[] { b };
+        c[0][0].ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         b[0].ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b[0]").WithLocation(8, 9),
+                // (10,9): warning CS8602: Possible dereference of a null reference.
+                //         c[0][0].ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "c[0][0]").WithLocation(10, 9));
+        }
+
+        [Fact]
+        public void ImplicitlyTypedArrayCreation_02()
+        {
+            var source =
+@"class C
+{
+    static void F(object x, object? y)
+    {
+        var a = new[] { x };
+        a[0].ToString();
+        var b = new[] { y };
+        b[0].ToString();
+        var c = new[] { a, b };
+        c[0][0].ToString();
+    }
+}";
+            var comp = CreateStandardCompilation(source, parseOptions: TestOptions.Regular8);
+            comp.VerifyDiagnostics(
+                // (8,9): warning CS8602: Possible dereference of a null reference.
+                //         b[0].ToString();
+                Diagnostic(ErrorCode.WRN_NullReferenceReceiver, "b[0]").WithLocation(8, 9),
+                // (9,28): warning CS8619: Nullability of reference types in value of type 'object?[]' doesn't match target type 'object[]'.
+                //         var c = new[] { a, b };
+                Diagnostic(ErrorCode.WRN_NullabilityMismatchInAssignment, "b").WithArguments("object?[]", "object[]").WithLocation(9, 28));
```

Comment:
Is this warning final, or do we expect it to change after conversion work?
I'd ultimately expect the same warning as previous test (warning on `c[0][0].ToString()` and `c` is found to be declared as array of nullables).

Reply:
This is not the final warning. This should be a conversion warning.

---
In reply to: [157622553](https://github.com/dotnet/roslyn/pull/23439#discussion_r157622553) [](ancestors = 157622553)